plugins/bnr-dp/fu-bnr-dp-device.c

```diff
@@ -0,0 +1,746 @@
+	g_autoptr(FuStructBnrDpAuxStatus) status = NULL;
+
+	/* only read the first 2 bytes of the header to check status bits. */
+	r = fu_dpaux_device_read(FU_DPAUX_DEVICE(self),
+				 FU_BNR_DP_DEVICE_HEADER_OFFSET,
+				 buf,
+				 sizeof(buf),
+				 FU_BNR_DP_DEVICE_DPAUX_TIMEOUT_MSEC,
+				 error);
+	if (!r)
+		return NULL;
+
+	status = fu_struct_bnr_dp_aux_status_parse(buf, sizeof(buf), 0, error);
+	if (status == NULL)
+		return NULL;
+
+	return g_steal_pointer(&status);
+}
+
+static gboolean
+fu_bnr_dp_device_poll_status(FuBnrDpDevice *self, GError **error)
+{
+	for (gint i = 0;; i++) {
+		g_autoptr(FuStructBnrDpAuxStatus) status = NULL;
+
+		status = fu_bnr_dp_device_read_status(self, error);
+		if (status == NULL)
+			return FALSE;
+
+		if (!fu_bnr_dp_device_eval_result(status, error))
+			return FALSE;
+		if (!fu_bnr_dp_device_is_busy(status))
+			return TRUE;
+
+		if (i > FU_BNR_DP_DEVICE_POLL_MAX_TOTAL) {
+			g_set_error_literal(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_BUSY,
+					    "polling device reached timeout");
+			return FALSE;
+		}
+		if (i > FU_BNR_DP_DEVICE_POLL_MAX_FAST)
+			g_usleep(FU_BNR_DP_DEVICE_POLL_INTERVAL_USEC);
+	}
+}
+
+static GByteArray *
+fu_bnr_dp_device_exec_cmd(FuBnrDpDevice *self,
+			  FuBnrDpOpcodes opcode,
+			  FuBnrDpModuleNumber module_number,
+			  guint16 offset,
+			  GError **error)
+{
+	g_autoptr(GByteArray) r = g_byte_array_new();
+	g_autoptr(FuStructBnrDpAuxRequest) request = NULL;
+
+	request = fu_bnr_dp_device_build_request(opcode, module_number, offset, 0, error);
+	if (request == NULL)
+		return NULL;
+
+	if (!fu_bnr_dp_device_write_request(self, request, NULL, 0, error))
+		return NULL;
+
+	if (!fu_bnr_dp_device_poll_status(self, error)) {
+		g_prefix_error(error,
+			       "command %s to module %s at offset 0x%X: ",
+			       fu_bnr_dp_opcodes_to_string(opcode),
+			       fu_bnr_dp_module_number_to_string(module_number),
+			       offset);
+		return NULL;
+	}
+
+	if (!fu_bnr_dp_device_read_response(self, r, error))
+		return NULL;
+
+	return g_steal_pointer(&r);
+}
+
+static GByteArray *
+fu_bnr_dp_device_read_data(FuBnrDpDevice *self,
+			   FuBnrDpOpcodes opcode,
+			   FuBnrDpModuleNumber module_number,
+			   gsize offset,
+			   gsize size,
+			   FuProgress *progress,
+			   GError **error)
+{
+	g_autoptr(GByteArray) r = g_byte_array_sized_new(size);
+	const guint16 start = offset / FU_BNR_DP_DEVICE_DATA_CHUNK_SIZE;
+	const guint16 end = (offset + size) / FU_BNR_DP_DEVICE_DATA_CHUNK_SIZE;
+
+	g_assert_cmpint(offset % FU_BNR_DP_DEVICE_DATA_CHUNK_SIZE, ==, 0);
+	g_assert_cmpint(size % FU_BNR_DP_DEVICE_DATA_CHUNK_SIZE, ==, 0);
+
+	for (guint16 idx = start; idx < end; idx++) {
+		g_autoptr(GByteArray) chunk = NULL;
+
+		chunk = fu_bnr_dp_device_exec_cmd(self, opcode, module_number, idx, error);
+		if (chunk == NULL)
+			return NULL;
+		g_byte_array_append(r, chunk->data, chunk->len);
+
+		fu_progress_set_percentage_full(progress, idx - start, end - start);
```

Comment:
Can you do fu_progress_set_steps() before the for() loop, and then use fu_progress_step_done() here? Then fwupd knows how many step's are remaining.